### ISSUE TITLE
Cannot style dialog content

### ISSUE THREAD

**Issue by heruan** [vaadin/vaadin-confirm-dialog-flow#38]
Styles applied to dialog content have no effect, as it seems the dialog content is rendered *inside* the `vaadin-dialog-overlay` shadow root.

**Comment by heruan** [comment]
Plus, any global `custom-style` or imported global modules (e.g. `lumo-typography`) have no effect to the dialog, which does not apply these styles.

**Comment by heruan** [comment]
Worth noting that styling and sizing works well with `Dialog`, but doesn't with `ConfirmDialog`. One would expect `ConfimDialog` be a specialization of `Dialog`, but I understand they are completely unrelated codebases.

**Comment by heruan** [comment]
More than a year has passed without any feedback. Some thoughts now? This is very limiting the use of the dialog, not being able to style its content 🙁

**Comment by heruan** [comment]
Thank you @javier-godoy for the PR which is useful, but this issue is not about the `theme` attribute or styling of the dialog: it's about the styling of the dialog *content*. For example:

```java
var colorful = new Span("My colorful content!");
colorful.addClassName("colorful");
var dialog = new ConfirmDialog();
dialog.add(colorful);
```

Since the `<span class="colorful">` is somehow rendered as part of the dialog overlay Shadow DOM (despite being added the Light DOM), the class `colorful` has no effect.

**Comment by jouni** [comment]
I agree that this is very limiting and breaks the consistent styling of an app.

What would need to change is that the default slot contents in the template are actually rendered in the light DOM. Needs a bit of thought how that plays together with content provided by the developer (some slotchange listeners and flags to mark which elements are provided by the component itself, perhaps).

**Comment by web-padawan** [comment]
This happens because of the fact that child nodes are being ["teleported"](https://github.com/vaadin/vaadin-confirm-dialog/blob/f1102e6eea9718e018cfc98953da994e1afb08e4/src/vaadin-confirm-dialog.html#L236-L237) separately.
I'm not sure whether trying to change that would be a breaking change or not.

**Comment by jouni** [comment]
Oh, true, there’s the teleporting as well.

I think it is indeed a breaking change if the content starts appearing in the light DOM.

Does anyone remember what was the reason to wrap `<vaadin-dialog>` instead of extend it?

**Comment by DiegoCardoso** [comment]
By having the `theme` propagated to the `vaadin-dialog`, doesn't that mean that the content can still be stylled with a style module to the `vaadin-dialog-overlay` with something like `:host([theme~="my-theme"]) .colorful {...}`?

It's not ideal, though.

**Comment by heruan** [comment]
> I think it is indeed a breaking change if the content starts appearing in the light DOM.

What this would actually break? Elements not getting styled is a bug that this would fix, not break; or there are other side-effects I'm not aware of?

> Does anyone remember what was the reason to wrap `<vaadin-dialog>` instead of extend it?

I really hope somebody does 😄 

> By having the `theme` propagated to the `vaadin-dialog`, doesn't that mean that the content can still be stylled with a style module to the `vaadin-dialog-overlay` with something like `:host([theme~="my-theme"]) .colorful {...}`?

@DiegoCardoso Yeah, sure but it's unpracticable since you'd need a theme module for *any* style for *all* dialogs; doable in a demo app maybe, not a real one.

**Comment by jouni** [comment]
> What this would actually break? Elements not getting styled is a bug that this would fix, not break; or there are other side-effects I'm not aware of?

Some apps might currently rely on the fact that styles do not match the contents of a confirm dialog, knowingly or unknowingly.

**Comment by heruan** [comment]
Well, fixing a bug is always a breaking change for those relying on the bug's side-effects 😉 But I understand this is borderline: if this won't land in V14 I root for V15 🙂

**Comment by Haprog** [comment]
I looked into this a bit to see what you can currently do regarding styling the internal parts of confirm dialog and styling your own content that you have put into the dialog.

You can style `vaadin-confirm-dialog` contents by making a style module with `theme-for="vaadin-dialog-overlay"`. This will of course affect all `vaadin-dialog-overlay` elements so once vaadin/vaadin-confirm-dialog#92 is merged, you can also scope your CSS to specific confirm dialogs by giving the confirm dialog element a custom `theme`.

@heruan your example:
```java
var colorful = new Span("My colorful content!");
colorful.addClassName("colorful");
var dialog = new ConfirmDialog();
dialog.add(colorful);
```

Can currently be styled just by adding:
```java
@CssImport(value = "./styles/dialog-overlay.css", themeFor = "vaadin-dialog-overlay")
```

Where `dialog-overlay.css` contains:
```css
.colorful {
    color: slateblue; /* or whatever styles you want */
}
```

### How it works

- User defined content is automatically moved to shadow root of `<vaadin-dialog-overlay>`
  - This is themable via `theme-for="vaadin-dialog-overlay"`
  - Theme attribute propagation would help this slightly to be able to target specific confirm dialog (vaadin/vaadin-confirm-dialog#92)
  - Is you need even more control over your content styles you could also create a new custom element just for wrapping your content in it so you can a add (and include) what styles you want into the template of your content wrapper element. I'm not sure how useful this is, but it would at least allow you to use `<style include="some-style-modules">` for your content without affecting all dialogs (would affect only those where you use your wrapper element).
- Dialog overlay content parts (`part="header"`, `part="message"`, `part="footer"`) are themable via `theme-for="vaadin-confirm-dialog"` because vaadin overlay automatically injects the styles from the host dialog into the shadow root of the content div (`<div part="content" id="content">`)
- When setting the contents via the `message` property of the web component (as a simple string) or as a Java `String` in the constructor or via `setText()` (instead of setting it as `Element` or `Component`) it gets added directly to the shadow root of the content div (instead of slotting them in via light dom) as a text node. In this case you can style this content by styling the `[part="message"]`.

### A styling example (as a Polymer 2 app)

This has two otherwise identical confirm dialogs except the second one has a custom theme which is used for targeting some styles only for that dialog. (This only works with the theme propagation fix mentioned above.)

```html
<link rel="import" href="../../bower_components/polymer/polymer-element.html">
<link rel="import" href="../../bower_components/vaadin-confirm-dialog/vaadin-confirm-dialog.html">
<link rel="import" href="../../bower_components/vaadin-lumo-styles/typography.html">

<dom-module id="custom-dialog-internal-styles" theme-for="vaadin-confirm-dialog">
    <template>
      <style include="lumo-typography">
        /*
        <div part="content" id="content"> element has a shadow root and inherits
        the vaadin-confirm-dialog styles

        :host selector styles can't be used here (for targeting the content div) since vaadin-overlay
        intentionally replaces ":host" with ":host-nomatch" to prevent host styles meant for
        vaadin-confirm-dialog element from matching the <div part="content" id="content"> element
        */

        /* style for vaadin-dialog-overlay content parts within all vaadin-confirm-dialogs */
        [part="header"] {
          color: blueviolet;
        }
        [part="message"] {
          font-size: 1.4rem;
        }

        /* When using the ConfirmDialog Flow component and setting the content as a String the text is added via the `message` property of the web component and it gets added directly to the shadow root of the content div (instead of slotting them in via light dom) as a text node. In this case you can style this content by styling [part="message"] here. */
      </style>
    </template>
  </dom-module>

<dom-module id="custom-dialog-content-styles" theme-for="vaadin-dialog-overlay">
  <template>
    <style include="lumo-typography">
      /* When using the vaadin-confirm-dialog you can target your own content elements in these styles. */

      /* style for all vaadin-dialog-overlays */
      p {
        border: 1px solid goldenrod;
      }

      /* style for vaadin-dialog-overlays with theme="custom-dialog"
      works with vaadin-confirm-dialog only after theme propagation fix is merged */
      :host([theme~="custom-dialog"]) .foo {
        font-weight: bold;
        background: coral;
      }
    </style>
  </template>
</dom-module>

<dom-module id="dialog-styling-app">
  <template>
    <vaadin-confirm-dialog id="dialog" header="Dialog" confirm-text="OK">
      <p>Lorem ipsum</p>
      <div class="foo">A block</div>
    </vaadin-confirm-dialog>

    <button on-click="openDialog">Open dialog 1</button>
    
    <vaadin-confirm-dialog id="dialog2" header="Dialog 2" confirm-text="OK" theme="custom-dialog">
      <p>Lorem ipsum</p>
      <div class="foo">A block</div>
    </vaadin-confirm-dialog>

    <button on-click="openDialog2">Open dialog 2 (custom theme)</button>
  </template>

  <script>
    class DialogStylingApp extends Polymer.Element {
      static get is() { return 'dialog-styling-app'; }
      openDialog() { this.$.dialog.opened = true; }
      openDialog2() { this.$.dialog2.opened = true; }
    }

    window.customElements.define(DialogStylingApp.is, DialogStylingApp);
  </script>
</dom-module>
```

### Conclusion

To me it looks like the confirm dialog contents can be styled pretty easily (when you know how) atm. But the biggest issue is that we don't seem to have any examples on how to do this . We should add styling examples to both web component demos and Flow component demos of confirm dialog (should open new issues about adding these styling demos, one issue in this repo and another in the web component repo).

I would like to close this issue as not an issue (can be styled already). But if you have some more specific use case which is still not possible or practical to style (considering above examples) it would probably be better to open a new more specific issue about those cases.

**Comment by jouni** [comment]
Thanks, Kari, for the through investigation and report!

The biggets gripe I have with this is that Dialog and Confirm Dialog need to be styled differently, AFAIK. If I place a `<h3>` inside a regular Dialog it will inherit styling from the global scope (assuming the dialog is placed in the global scope), but with Confirm Dialog I need to explicitly include the global typography styles with an extra style module for vaadin-dialog-overlay.

**Comment by heruan** [comment]
Thank you @Haprog for the detailed overview and suggestions! I just think the issue turned out to be a fundamental maintainability one: you need to create theme modules to style content you added to the light DOM, content that could be a component defined in an external dependency (e.g. from Vaadin Directory).

I agree that "confirm dialog contents can be styled", but currently it requires to take special care where it shouldn't be necessary.

**Comment by heruan** [comment]
Any update on this? Light DOM elements still aren't styled as they should (because they're moved to the Shadow DOM). I hardly see how this is not a bug 😕

**Comment by jouni** [comment]
Not sure if fixing a bug can be considered a breaking change? I suppose at this point, when the behavior has been stable for so long, it can be, and we would need to do the change in a new major version. Any thoughts on that?

**Comment by Haprog** [comment]
I'm closing this as a non-issue ("Cannot style dialog content"). See my previous comment above https://github.com/vaadin/vaadin-confirm-dialog-flow/issues/38#issuecomment-548741322

Afaik it works as intended (when implemented) and styling is "possible" (though not convenient). Also it seems like this can't be made any better in a minor or patch version.

@heruan please open a new issue with a clear description specifically about not moving the Light DOM elements (so that they can inherit styles in a normal way) if that is what you want. I think it's better to start discussing that in a new issue and it could maybe be fixed for a new major version at some point.

**Comment by heruan** [comment]
Since the workaround provided here isn't feasible when there are many dialogs with dynamic content (i.e. one cannot use themes to style them) I've opened this:

https://github.com/vaadin/vaadin-confirm-dialog/issues/106
